File: platform/core/src/core_failure.rs
use serde::Serialize;
use url::Url;

#[cfg(target_os = "android")]
use crate::rdf_failure::RdfFailure;

#[derive(Debug, Clone, Serialize)]
pub enum FailureCode {
    CoreNotBootstrapped = 1,
    CoreAlreadyBootstrapped,
    FailedToParseFeatureManifest,
    NullCStringPointer,
    FailedToCreateCString,
    FailedToExtractJavaString,
    FailedToConvertJavaString,
    FailedToParseFeatureCategoriesJSON,
    FailedToReadFile,
    FailedFileSystemOperation,
    FailedToParseURL,
    FailedToUnzip,
    FailedToCreateFeatureFilesPath,
    FailedToConvertToFsPath,
    FailedToConvertToResourceUrl,
    FailedToGetFilePath,
    FailedToGetLastSegmentFromUrl,
}

impl FailureCode {
    fn value(&self) -> i32 {
        (*self).clone() as i32
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct CoreFailure {
    pub code: i32,
    pub message: String,
}

impl CoreFailure {
    pub fn core_already_bootstrapped() -> Self {
        CoreFailure {
            code: FailureCode::CoreAlreadyBootstrapped.value(),
            message: "Core was already initialized".to_string(),
        }
    }

    pub fn core_not_bootstrapped() -> Self {
        CoreFailure {
            code: FailureCode::CoreNotBootstrapped.value(),
            message: "Core was not initialized".to_string(),
        }
    }

    pub fn failed_to_parse_feature_manifest(message: String) -> Self {
        CoreFailure {
            code: FailureCode::FailedToParseFeatureManifest.value(),
            message,
        }
    }

    pub fn null_c_string_pointer() -> Self {
        CoreFailure {
            code: FailureCode::NullCStringPointer.value(),
            message: "cstring pointer is null.".to_string(),
        }
    }

    pub fn failed_to_create_c_str(message: String) -> Self {
        CoreFailure {
            code: FailureCode::FailedToCreateCString.value(),
            message,
        }
    }

    #[cfg(target_os = "android")]
    pub fn map_rdf_to_core_failure(error: RdfFailure) -> Self {
        CoreFailure {
            code: error.code,
            message: error.message 
        }
    }

    #[cfg(target_os = "android")]
    pub fn failed_to_extract_java_string(message: String) -> Self {
        CoreFailure {
            code: FailureCode::FailedToExtractJavaString.value(),
            message,
        }
    }

    #[cfg(target_os = "android")]
    pub fn failed_to_convert_java_string(message: String) -> Self {
        CoreFailure {
            code: FailureCode::FailedToConvertJavaString.value(),
            message,
        }
    }

    pub fn failed_to_decode_feature_categories_json(message: String) -> Self {
        CoreFailure {
            code: FailureCode::FailedToParseFeatureCategoriesJSON.value(),
            message,
        }
    }

    pub fn failed_to_read_contents_of_file(path: String, message: String) -> Self {
        CoreFailure {
            code: FailureCode::FailedToReadFile.value(),
            message: format!(
                "Failed to read file from path '{}', error: '{}'",
                path, message,
            ),
        }
    }

    pub fn failed_file_system_operation(path: String, message: String) -> Self {
        CoreFailure {
            code: FailureCode::FailedFileSystemOperation.value(),
            message: format!(
                "File system failed for path '{}' with error: '{}'",
                path, message,
            ),
        }
    }

    pub fn failed_to_parse_url(url: String, message: String) -> Self {
        CoreFailure {
            code: FailureCode::FailedToParseURL.value(),
            message: format!("Failed to parse url '{}' with error: '{}'", url, message,),
        }
    }

    pub fn failed_to_unzip(url: String, message: String) -> Self {
        CoreFailure {
            code: FailureCode::FailedToUnzip.value(),
            message: format!(
                "Failed to unzip resource at url '{}' with error: '{}'",
                url, message,
            ),
        }
    }

    pub fn failed_to_create_feature_files_path(message: String) -> Self {
        CoreFailure {
            code: FailureCode::FailedToCreateFeatureFilesPath.value(),
            message: format!(
                "Failed to create feature files path with error: '{}'",
                message,
            ),
        }
    }

    pub fn failed_to_convert_to_fs_path_from_resource_url(url: String, message: String) -> Self {
        CoreFailure {
            code: FailureCode::FailedToConvertToFsPath.value(),
            message: format!(
                "Failed to create fs path from resource url {} with error: '{}'",
                url, message,
            ),
        }
    }

    pub fn failed_to_convert_to_resource_url_from_fs_path(url: String, message: String) -> Self {
        CoreFailure {
            code: FailureCode::FailedToConvertToResourceUrl.value(),
            message: format!(
                "Failed to create fs path from resource url {} with error: '{}'",
                url, message,
            ),
        }
    }

    pub fn failed_to_get_file_path(url: Url, message: String) -> Self {
        CoreFailure {
            code: FailureCode::FailedToGetFilePath.value(),
            message: format!(
                "Failed to get file path from url {} with error: '{}'",
                url, message,
            ),
        }
    }

    pub fn failed_to_get_last_segment_from_url(url: Url, message: String) -> Self {
        CoreFailure {
            code: FailureCode::FailedToGetLastSegmentFromUrl.value(),
            message: format!(
                "Failed to get last segment from url {} with error: '{}'",
                url, message,
            ),
        }
    }
}

File: platform/core/PolyPodCoreSwift/Sources/PolyPodCoreSwift/CoreFailure.swift
import MessagePack
public enum DecodingError: Error {
    case invalidValue(info: String)
    case invalidCoreFailure(info: String)
    case invalidCoreResult(info: String)
    case missingDictionaryKey(info: String)
    case invalidValueType(info: String)
    case invalidResponse(info: String?)
    case invalidFeatureCategoryFormat
    case unknownFeatureCategoryId(info: String)
    case unknownUserSessionTimeoutOption(info: String)
    
    var localizedDescription: String {
        switch self {
        case .invalidValue(info: let info):
            return "Received invalid message pack value: \(info)"
        case .invalidCoreFailure(info: let info):
            return "Invalid core failure object format \(info)"
        case .invalidCoreResult(info: let info):
            return "Invalid core result object format \(info)"
        case .missingDictionaryKey(info: let info):
            return "Missing key in dictionary object: \(info)"
        case .invalidValueType(info: let info):
            return "Invalid value type: \(info)"
        case let .invalidResponse(info):
            return "Received invalid core response: \(String(describing: info))"
        case .invalidFeatureCategoryFormat:
            return "Expected dictionary for feature category"
        case let .unknownFeatureCategoryId(info):
            return "Unknown FeatureCategoryId: \(info)"
        case let .unknownUserSessionTimeoutOption(info):
            return "Unknown UserSessionTimeoutOption: \(info)"
        }
    }
}

public enum CoreFailureCode: Int, Codable {
    case coreNotBootstrapped = 1
    case coreAlreadyBootstrapped
    case failedToParseFeatureManifest
    case nullCStringPointer
    case failedToCreateCString
    case failedToExtractJavaString
    case failedToConvertJavaString
    case failedToParseFeatureCategoriesJSON
    case failedToReadFile
    case failedFileSystemOperation
    case failedToParseURL
    case failedToUnzip
    case failedToCreateFeatureFilesPath
    case failedToConvertToFsPath
    case failedToConvertToResourceUrl
    case failedToGetFilePath
    case failedToGetLastSegmentFromUrl
    case failedToDecodeByteArray
    case failedToReadByteBufferLength
    case failedToExtractJObject
    case failedToExtractBytes
    case failedToCallJNIMethod
    case failedToConvertBytes
    case failedToAccessUserSession
    case failedToAttachJVM
    case failedToDecode
    case failedToEncode
    case failedToSetPreference
    
    init(from value: MessagePackValue) throws {
        guard let code = CoreFailureCode.init(rawValue: try value.getInt()) else {
            throw DecodingError.invalidCoreFailure(info: "Received \(value)")
        }
        self = code
    }
}

public struct CoreFailure: Error, Codable {
    public let code: CoreFailureCode
    public let message: String
    
    public init(code: CoreFailureCode, message: String) {
        self.code = code
        self.message = message
    }
    
    init(from value: MessagePackValue) throws {
        let dict: [MessagePackValue: MessagePackValue] = try value.getDictionary()
        code = try CoreFailureCode(from: dict.get("code"))
        message = try dict.get("message").getString()
    }
}
